src/v/kafka/client/client.cc

```diff
@@ -341,30 +345,72 @@ client::create_topic(kafka::creatable_topic req) {
     });
 }
 
-ss::future<list_offsets_response>
-client::list_offsets(model::topic_partition tp) {
-    return gated_retry_with_mitigation(
-             [this, tp{std::move(tp)}]() { return do_list_offsets(tp); })
-      .then(&throw_on_error<list_offsets_response>);
+ss::future<list_offsets_response> client::list_offsets(
+  chunked_vector<model::topic_partition> tps, model::timestamp timestamp) {
+    co_return co_await gated_retry_with_mitigation([this, &tps, timestamp]() {
+        return do_list_offsets(tps.copy(), timestamp);
+    }).then(&throw_on_error<list_offsets_response>);
 }
 
-ss::future<list_offsets_response>
-client::do_list_offsets(model::topic_partition tp) {
-    auto node_id = co_await _topic_cache.leader(tp);
-    auto broker = co_await _brokers.find(node_id);
-    chunked_vector<kafka::list_offset_topic> cv;
-    cv.push_back(kafka::list_offset_topic{
-      .name{tp.topic},
-      .partitions{
-        {
-          {.partition_index{tp.partition}, .max_num_offsets = 1},
-        },
-      },
-    });
-    auto res = co_await broker->dispatch(kafka::list_offsets_request{
-      .data = {
-        .topics = std::move(cv),
-      }});
+ss::future<list_offsets_response> client::do_list_offsets(
+  chunked_vector<model::topic_partition> tps, model::timestamp timestamp) {
+    // Group requests by the broker they are sent to
+    chunked_hash_map<model::node_id, kafka::list_offsets_request> reqs;
+    for (const auto& tp : tps) {
+        auto node_id = co_await _topic_cache.leader(tp);
+
+        auto [req, inserted] = reqs.emplace(
+          node_id, kafka::list_offsets_request{});
+        auto& topics = req->second.data.topics;
+        auto topic_it = std::ranges::find_if(
+          topics, [&](auto& top) { return top.name == tp.topic; });
+        auto make_partition = [&]() {
+            return kafka::list_offset_partition{
+              .partition_index{tp.partition},
+              .timestamp = timestamp,
+              .max_num_offsets = 1,
+            };
+        };
+        if (topic_it != topics.end()) {
+            topic_it->partitions.emplace_back(make_partition());
+        } else {
+            topics.emplace_back(kafka::list_offset_topic{
+              .name{tp.topic}, .partitions = {make_partition()}});
+        }
+    }
```

Comment:
This could be a little syntactically less verbose, the topic can be moved, and default construction can be avoided if the item already exists:
```suggestion
    for (auto& tp : tps) {
        auto node_id = co_await _topic_cache.leader(tp);
        auto& topics = reqs.try_emplace(node_id).first->second.data.topics;
        auto topic_it = std::ranges::find(
          topics, tp.topic, &list_offset_topic::name);
        auto& topic = topic_it == topics.end()
                        ? topics.emplace_back(std::move(tp.topic))
                        : *topic_it;
        topic.partitions.emplace_back(
          tp.partition, kafka::leader_epoch{}, timestamp);
    }
```